### PR TITLE
Use bloom filter to optimize df.merge execution

### PR DESCRIPTION
**Pull request by hekaisheng**
## What do these changes do?

Use small DataFrame to build a bloom filter and filter the other DataFrame before merge, it will extremely reduce the number of data transfers when shuffling them. 

It's now a rough implementation, some details need to be confirmed.

## Related issue number

Fixes #xxxx

## Check code requirements

- [x] tests added / passed (if needed)
- [x] Ensure all linting tests pass, see [here](https://docs.pymars.org/en/latest/development/contributing.html#check-code-styles) for how to run them

### CODE REVIEW

File: mars/dataframe/base/bloom_filter.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+        return bloom_filter
+
+    @classmethod
+    def _build_dataframe_filter(cls, in_data, op):
+        try:
+            bloom_filter = BloomFilter(
+                max_elements=op.max_elements, error_rate=op.error_rate
+            )
+            in_data[op.on].apply(lambda v: bloom_filter.add(cls._get_value(v)), axis=1)
+        except TypeError:  # pragma: no cover
```
Comment (qinxuye) [inline mars/dataframe/base/bloom_filter.py:104]: Better to add tests instead of skipping.

File: mars/dataframe/merge/merge.py [hunk 2]
```diff
@@ -904,6 +982,9 @@ def join(
         When how is "inner", merged result could be much smaller than original DataFrame,
         if the number of chunks is greater than the threshold,
         it will merge small chunks automatically.
+    bloom_filter: bool or dict, default True
```
Comment (qinxuye) [inline mars/dataframe/merge/merge.py:985]: Add doc about if dict, which keys can have.

**Review by qinxuye — COMMENTED** [review]
Coverage can still be improved.

File: mars/dataframe/merge/merge.py (hunk 2 above)
Comment (qinxuye) [reply mars/dataframe/merge/merge.py:985]: Besides keys, the default values can be mentioned.

File: mars/dataframe/base/bloom_filter.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+    def __init__(self, execution_stage=None, **kwargs):
+        if execution_stage in ["build", "union"]:
+            output_types = [OutputType.object]
+        else:
+            output_types = [OutputType.dataframe]
+        kwargs["_output_types"] = output_types
+        super().__init__(execution_stage=execution_stage, **kwargs)
+
+    @classmethod
+    def _get_value(cls, value):
```
[the remark this answers is not in the document]
Comment (qinxuye) [reply mars/dataframe/base/bloom_filter.py:52]: Still no type annotation, so does the other functions like `_filter_on_index` etc.

File: mars/dataframe/base/bloom_filter.py [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+                            cls._convert_to_hashable_dtypes(in_data.dtypes)
+                        )
+                        filtered = in_data[converted_data[on].apply(row_func, axis=1)]
+                    ctx[op.outputs[0].key] = filtered
+
+        else:  # pragma: no cover
+            raise ValueError(f"Unknown execution stage: {op.execution_stage}")
+
+
+def filter_by_bloom_filter(
```
Comment (qinxuye) [inline mars/dataframe/base/bloom_filter.py:181]: Maybe this function could be tileable level, and the chunks related stuff could be moved to the `tile` function, I guess this function could be exposed to user some day perhaps, not just be taken as a utility function.

File: mars/dataframe/merge/merge.py [hunk 5]
```diff
@@ -245,6 +255,32 @@ def _gen_shuffle_chunks(
             )
         return reduce_chunks
 
+    @classmethod
+    def _gen_bloom_filter_chunks(cls, left, right, left_on, right_on, op):
```
Comment (qinxuye) [inline mars/dataframe/merge/merge.py:259]: Type annotation, and seems like this function returns two tileables instead of `chunks` in the function name.

**Review by qinxuye — COMMENTED** [review]
All new functions should be typed annotated, besides, I left some other comments.

**Review by qinxuye — COMMENTED** [review]
All new functions should be typed annotated, besides, I left some other comments.

File: mars/dataframe/base/bloom_filter.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hekaisheng) [reply mars/dataframe/base/bloom_filter.py:52]: value could be a scalar like string, int, float, etc. I have added comments about the value.

**Review by qinxuye — APPROVED** [review]
LGTM

**Review by wjsi — APPROVED** [review]
LGTM

**Pull request merged**